salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python2
+
+from __future__ import print_function
+import argparse
+import errno
+import os
+import stat
+import sys
+
+
+SPARSE_FILES_DIR = '/var/lib/metalk8s/storage/sparse/'
+LOOP_MAJOR = 7
+
+
+def parse_args(args=None):
+    parser = argparse.ArgumentParser(
+        prog='metalk8s-sparse-volume-cleanup',
+        help='Cleanup MetalK8s sparse loop volumes',
+    )
+
+    parser.add_argument(
+        '-i', '--uid',
+        required=True,
+        help='UID of the MetalK8s Volume object',
+        dest='volume_id',
+    )
+
+    return parser.parse_args(args=args)
+
+
+def check_sparse_file(volume_id):
+    """Check if the sparse file exists for this volume ID."""
+    sparse_file_path = SPARSE_FILES_DIR + volume_id
+    if not os.path.isfile(sparse_file_path):
+        die(
+            "Sparse file {} does not exist for volume '{}'".format(
+                sparse_file_path, volume_id
+            ),
+            exit_code=errno.ENOENT
+        )
+
+
+def cleanup(volume_id):
+    device_path = find_device(volume_id)
+
+    print("Detaching loop device '{}' for volume '{}'".format(
+        device_path, volume_id
+    ))
+
+    fd = os.open(device_path, os.O_RDONLY)
+    try:
+        fcntl.ioctl(fd, LOOP_CLR_FD, 0)
+    except IOError as exn:
+        if exn.errno != errno.ENXIO:
+            die("Unexpected error when trying to free device {}: {}".format(
+                device_path, str(exn)
+            ))
+        print("Device already freed", file=sys.stderr)
+    finally:
+        fd.close()
+
+    print("Loop device for volume '{}' was successfully detached".format(
+        volume_id
+    ))
+
+
+def main():
+    args = parse_args()
+
+    check_sparse_file(args.volume_id)
+    cleanup(args.volume_id)
+
+
+# Helpers {{{
+def die(message, exit_code=1):
+    print(message, file=sys.stderr)
+    sys.exit(exit_code)
+
+
+def find_device(volume_id):
+    """Find the device provisioned for this volume ID.
+    
+    SparseLoopDevice volumes are either:
+    - formatted, with the volume ID set in the filesystem annotations,
+      hence discoverable under /dev/disk/by-uuid/
+    - raw, in which case the device is partitioned and the first partition
+      is labeled with the volume ID, hence discoverable under
+      /dev/disk/by-partuuid/
+    """
+    path_by_uuid = "/dev/disk/by-uuid/{}".format(volume_id)
+    if os.path.exists(path_by_uuid):
+        device_path = os.path.realpath(path_by_uuid)
+    else:
+        path_by_partuuid = "/dev/disk/by-partuuid/{}".format(volume_id)
+        if not os.path.exists(path_by_partuuid):
```

Comment:
Now if we'd add a third type of device (discovery), this would become one more indented block, and so on. I'd put all 'discovery' code (per case) in a separate function, then call them one by one until one returns a non-`None` result (or doesn't raise, whatever), then fail if none of them return a 'good' value: cleaner code, better scoped, less nesting and easier to test.